File: 00-Foundations/04-Design-Patterns/Creational/Builder/Builder.cs
namespace Foundations.Patterns.Creational.Builder;

// The finished product is immutable — no setters, only a constructor the
// builder calls once it has gathered everything.
public sealed class Pizza
{
    public string Size { get; }
    public IReadOnlyList<string> Toppings { get; }
    public bool ExtraCheese { get; }

    public Pizza(string size, IReadOnlyList<string> toppings, bool extraCheese)
    {
        Size = size;
        Toppings = toppings;
        ExtraCheese = extraCheese;
    }

    public override string ToString() =>
        $"{Size} pizza with [{string.Join(", ", Toppings)}]{(ExtraCheese ? " + extra cheese" : "")}";
}

public class PizzaBuilder
{
    private string _size = "Medium";
    private readonly List<string> _toppings = new();
    private bool _extraCheese;

    public PizzaBuilder WithSize(string size)
    {
        _size = size;
        return this; // fluent chaining
    }

    public PizzaBuilder AddTopping(string topping)
    {
        _toppings.Add(topping);
        return this;
    }

    public PizzaBuilder WithExtraCheese()
    {
        _extraCheese = true;
        return this;
    }

    public Pizza Build() => new(_size, _toppings, _extraCheese);
}

public static class BuilderDemo
{
    public static void Run()
    {
        Pizza pizza = new PizzaBuilder()
            .WithSize("Large")
            .AddTopping("Mushroom")
            .AddTopping("Olives")
            .WithExtraCheese()
            .Build();

        Console.WriteLine(pizza);
    }
}

File: 00-Foundations/04-Design-Patterns/Creational/Builder/builder.ts
// TypeScript often reaches for a plain object with optional properties
// instead of a Builder class (`function makePizza(opts: { size?: string, ... })`),
// since destructured optional params already solve "telescoping constructor."
// The class-based Builder still earns its keep when construction has *steps*
// with validation between them, or must stay fluent/chainable — shown here.

class Pizza {
  constructor(
    readonly size: string,
    readonly toppings: ReadonlyArray<string>,
    readonly extraCheese: boolean,
  ) {}

  toString(): string {
    return `${this.size} pizza with [${this.toppings.join(", ")}]${this.extraCheese ? " + extra cheese" : ""}`;
  }
}

class PizzaBuilder {
  private size = "Medium";
  private toppings: string[] = [];
  private extraCheese = false;

  withSize(size: string): this {
    this.size = size;
    return this;
  }

  addTopping(topping: string): this {
    this.toppings.push(topping);
    return this;
  }

  withExtraCheese(): this {
    this.extraCheese = true;
    return this;
  }

  build(): Pizza {
    return new Pizza(this.size, this.toppings, this.extraCheese);
  }
}

function demo(): void {
  const pizza = new PizzaBuilder()
    .withSize("Large")
    .addTopping("Mushroom")
    .addTopping("Olives")
    .withExtraCheese()
    .build();

  console.log(pizza.toString());
}

demo();
